docs: WIP mutable schemas (#967)
empathyco/x: packages/x-adapter/README.md

#### Using a mutable schema

This feature lets you have some default mappers, and modify or extend them for some concrete
implementations. To do so, you should use the `createMutableSchema` helper function, and pass as
parameters a `Source` and a `Target` to map your models, it will return a `MutableSchema` that will
accept a `NewSource` and a `NewTarget` to handle the transformations you need.

In the example below we will create a `MutableSchema` to have a default object that will be reused
for different endpoint calls.

###### Types definition and MutableSchema

```ts
// API models
export interface ApiBaseObject {
  id: number;
  body: string;
}

// APP models
export interface AppBaseObject {
  id: string;
  text: string;
}

// Mutable Schema
export const baseObjectSchema = createMutableSchema<ApiBaseObject, AppBaseObject>({
  id: ({ id }) => id.toString(),
  text: 'body'
});
```

> Once we have the `MutableSchema`, we can use the `createMutableSchema` available methods to fit our
> different APIs needs:

Reviewer: ```suggestion
Once we have the `MutableSchema`, we can use the following methods to fit our
different APIs needs:
```

> - `extends`: Creates a new `Schema` based on the original (default) one. The original remains

Reviewer: ```suggestion
- `$extends`: Creates a new `Schema` based on the original (default) one. The original remains
```

Author: Catched 👀

  unchanged. Useful if we need to create a new `EndpointAdapter` to fit a new API.
- `override`: Merges/modifies the original `Schema` partially, so the change will affect to all the
  `EndpointAdapter`(s) that are using it. It can be used to change the structure of our
  request/response mappers, or to add them new fields. Useful for clients with few differences in
  their APIs (so they can use an already created adapter and override it to make some tiny changes,
  e.g. change from 'image' field to 'images').

> - `replace`: Replaces completely the original `Schema` by a new one, it won't exist anymore. The

Reviewer: ```suggestion
- `$replace`: Replaces completely the original `Schema` by a new one, it won't exist anymore. The
```

});
```

###### Replace a MutableSchema to completely change it

In this case we are facing too many differences between API responses. We don't need to write a
whole adapter from scratch, as there are other parts of the API that aren't changing so much, but we
should replace some `endpointAdapter`'s schemas.

```ts
import { ApiBaseObject, AppBaseObject, baseObjectSchema } from '@/base-types';

// Api models
interface ApiQuote {
  id: number;
  quote: string;
  author: string;
}

interface ApiQuotesResponse {
  quotes: ApiBaseObject[];
}

// App models
interface AppQuote {
  quoteId: string;
  quote: string;
  author: string;
}

interface AppQuotesResponse {
  quotes: AppBaseObject[];
}

// Replace the original Schema
baseObjectSchema.$replace<ApiQuote, AppQuote>({
  quoteId: ({ id }) => id.toString(),
  quote: 'quote',
  author: 'author'
});

// Response mapper
const quotesResponse = schemaMapperFactory<ApiQuotesResponse, AppQuotesResponse>({
  quotes: {
    $subSchema: baseObjectSchema,
    $path: 'quotes'
  }
});

// Endpoint Adapter
export const searchQuotes = endpointAdapterFactory({
  endpoint: 'https://dummyjson.com/quotes',
  responseMapper: quotesResponse

> });

Reviewer: To emphasize that `$replace` and `$override` mutate the original schemas, and therefor the adapter, I would move the `$replace` and `$override` calls after the creation of the endpoint adapter.

#### Using a mutable schema

This feature lets you have some default mappers, and modify or extend them for some concrete
implementations. To do so, you can use the `createMutableSchema` function, passing a `Source` and
`Target` type parameters to map your models. This function will return a `MutableSchema` that apart
from the mapping information will also contain some methods to create new schemas or modify the
current one.

In the example below we will create a `MutableSchema` to have a default object that will be reused
for different endpoint calls.

###### Types definition and MutableSchema

```ts
// API models
export interface ApiBaseObject {
  id: number;
  body: string;
}

// APP models
export interface AppBaseObject {
  id: string;
  text: string;
}

// Mutable Schema
export const baseObjectSchema = createMutableSchema<ApiBaseObject, AppBaseObject>({
  id: ({ id }) => id.toString(),
  text: 'body'
});
```

Once we have the `MutableSchema`, we can use the following methods to fit our different APIs needs:

> - `$extends`: Creates a new `Schema` based on the original one. The original remains unchanged. This

Author: Good catch! 🚀

  can be useful if we need to create a new `EndpointAdapter` with models based on another API.
- `$override`: Merges/modifies the original `Schema` partially, so the change will affect to all the
  `EndpointAdapter`(s) that are using it. It can be used to change the structure of our
  request/response mappers, or to add them new fields. Useful for clients with few differences in
  their APIs. For example, you can create a library with a default adapter and use this library from
  the customer projects overriding only the needed field (e.g. retrieve the images from `pictures`
  instead of `images` in a products API).
- `$replace`: Replaces completely the original `Schema` by a new one, it won't exist anymore. The
  change will affect to all the `EndpointAdapter`(s) that were using it. Useful for clients with a
  completely different API/response to the standard you have been working with.

###### Extend a MutableSchema to reuse it in two different endpoints with more fields

```ts

> // Api models

Reviewer 2: ```suggestion
import { ApiBaseObject, AppBaseObject, baseObjectSchema } from '@/base-types';
// Api models
```
